Add free-hand drawing of sliders to the editor

Tom94 opened the pull request:
- [x] Depends on https://github.com/ppy/osu-framework/pull/6044
- [x] Depends on https://github.com/ppy/osu-framework/pull/6047
- [x] Depends on https://github.com/ppy/osu-framework/pull/6053

Collab with @default0.

This PR allows free-hand drawing of a slider by click+dragging in the osu!-mode editor. As a dependent addition, BSplines were integrated as a new slider type that generalize Bezier curves (corresponding to BSplines with degree equal to the number of control points). The new BSpline slider type is used to represent / approximate the hand-drawn slider and have extra logic for adaptive control point spacing in "pointy" slider regions.

https://github.com/ppy/osu/assets/4923655/7852d901-bdc7-4129-a8e3-714590fc76e5

Tom94 committed:
Generalize Bezier curves to BSplines of Nth degree

PM-CS committed:
Add free-hand drawing of sliders to the editor

PM-CS committed:
CI Fixup

bdach commented on osu.Game.Rulesets.Osu/Edit/OsuSliderDrawingSettingsProvider.cs:
```diff
@@ -0,0 +1,66 @@
+            });
+        }
+
+        public void AttachToToolbox(ExpandingToolboxContainer toolboxContainer)
```
We probably want to make a `IToolboxAttachment` interface out of this method or something given it's showed up on a second class now. But maybe as a follow-up PR.

Tom94 committed:
Fix OnDragStart on macOS

Tom94 committed:
Fix doubled control point at beginning of drawn slider

peppy committed:
Update framework

peppy commented:
In initial testing I have concerns over the usability of this feature.

- "C.P.S." higher than 20 or so it's pretty useless, so I'd limit it more
- Even with it lower, it's very hard to make shapes with corners (most of the time I try it ends up going wrong)
- It's very hard to use the freehand to create a shape you're trying to make.
- The resultant points are too frequent to provide a good manual tweaking experience after creating

I think if this was to be released, there will be a largely negative reaction from the modding scene as they deal with users making shitty sliders (and the amount of modding overhead to recreate sliders using proper nodes). We may want to wait for a better algorithm before releasing this to avoid ongoing negative connotations with a potentially cool feature.

Example 1: trying to match a slider to another slider's shape (i can't make it work):

https://github.com/ppy/osu/assets/191335/8691e888-c9d7-4477-a598-cc245b94c5e0

Example 2: corners just feel bad (and aren't creating a new curve segment):

https://github.com/ppy/osu/assets/191335/acdb0f87-eb11-46ae-b2e7-a06c1f5045b9

Example 3: control point spacing set to the max value, useless

https://github.com/ppy/osu/assets/191335/2b37e777-a5b3-44a6-ac58-f5c2c7b16598

Also:

- Control Point Spacing is a mouthful, dunno if we can come up with something better.
- I'd change the heading of the toolbox area from "drawing" to "slider".

peppy reviewed (commented):
Just a usability pass, haven't checked the code.

PM-CS committed:
Fix code style/quality issues

Tom94 committed:
Automatic red control point generation & corner threshold

PM-CS committed:
CI fixes and small tweaks

PM-CS committed:
Fix control point hover text and context menu

PM-CS committed:
Hide catmull curve type when possible

bdach committed:
Merge branch 'master' into bspline-sliders

bdach committed:
Update framework (again)

bdach committed:
Move static instances / construction methods closer together

bdach committed:
Change switchexpr to standard switch statement

bdach committed:
Decrease redundancy of equality implementations

bdach committed:
Null-check `drawingSettingsProvider`

As it's annotated as an optional dependency.

bdach committed:
Avoid one unnecessary path update from B-spline builder

bdach committed:
Remove unused enum member

bdach commented on osu.Game.Rulesets.Osu/Edit/ISliderDrawingSettingsProvider.cs:
This seems to be excessive abstraction to me, so I've removed it for now.

bdach committed:
Remove `ISliderDrawingSettingsProvider`

Seems like excessive abstraction.

bdach commented on osu.Game/Rulesets/Edit/IToolboxAttachment.cs:
I'm also removing this for now. Yes I suggested it, but (a) the interface doesn't really do anything useful right now because it enforces a common contract, but all usages of the contract go through the concrete implementations, and (b) it inflates the already-huge diff. So to have a meaningful code impact I'd say it should be done in a follow-up.

As a loose sketch, I'd imagine that `HitObjectComposer` should get an `Attach(IToolboxAttachment)` method that does all of the hooking up that needs to happen. It'll probably require a typecheck to `Drawable` to handle the `AddInternal()` part, but that's probably okay.

bdach committed:
Remove `IToolboxAttachment` for now

The interface doesn't really do anything useful right now because
it enforces a common contract, but all usages of the contract
go through the concrete implementation, and it inflates
the already-huge diff.

bdach committed:
Remove pattern matching syntax usage in switch

Also throw on unknown types.

bdach commented on osu.Game/Rulesets/Objects/BezierConverter.cs:
```diff
@@ -124,46 +124,46 @@ public static List<PathControlPoint> ConvertToModernBezier(IList<PathControlPoin
                         for (int j = 0; j < segmentVertices.Length - 1; j++)
                         {
-                            result.Add(new PathControlPoint(segmentVertices[j], j == 0 ? PathType.Bezier : null));
+                            result.Add(new PathControlPoint(segmentVertices[j], j == 0 ? segmentType : null));
```
This looks like it's going to preserve B-spline nodes of nonstandard degrees as they were. This is probably going to be an issue as the method being changed here is used to backport lazer sliders to stable, and stable cannot support the new `Bxx` curve specs.

I'm not yet sure how much of a problem this is going to be, and am yet to test, but yeah...

bdach replied:
Yeah this produces broken maps when exporting for compatibility to stable.

Thankfully, as far as I can tell, framework [already implements Boehm's algorithm](https://github.com/ppy/osu-framework/blob/2024a31b07e8896897312b79323b890f5b4e1583/osu.Framework/Utils/PathApproximator.cs#L84-L108) to remedy this, it just needs to be exposed for the game.

bdach committed:
Use static `BEZIER` instead of allocating new every time

bdach committed:
Fix incorrect legacy conversion when B-splines are used

bdach committed:
Add test for slider drawing

bdach committed:
Remove test terminally broken by introduction of slider drawing

bdach committed:
Adjust casing of curve type menu items

The "Perfect curve" one in particular... fixes test failures, as some
tests were relying on this particular casing. But the new version feels
more correct anyway, so it's whatever.

bdach committed:
Show Catmull as a control point type option if selection already contains it

bdach committed:
Decrease default value for slider tolerance

Highly subjective change, but at 50 the drawing just did not feel
responsive enough to input.

bdach commented:
I did stuff here. I'm pretty okay with the state of this after the stuff I did, but (a) it now requires https://github.com/ppy/osu-framework/pull/6053 and (b) I'll probably defer to @peppy when it comes to retesting the UX concerns mentioned above.

peppy committed:
Merge branch 'master' into bspline-sliders

peppy committed:
Update framework

peppy committed:
Remove unnecessary null override

peppy committed:
Simplify null checks

peppy committed:
Rename settings class

peppy committed:
Get rid of weird cruft and non-standard flow

peppy commented:
I really dunno. Is it just me or does using this in the editor feel shoddy still? I feel like it's not going to make a curve because when I'm drawing it's just a straight line to my mouse point. UX has to improve.

peppy committed:
Refactor slider settings class

peppy commented on osu.Game.Rulesets.Osu/Edit/FreehandSliderSettingsProvider.cs:
```diff
@@ -0,0 +1,108 @@
+namespace osu.Game.Rulesets.Osu.Edit
+{
+    public partial class FreehandSliderSettingsProvider : EditorToolboxGroup
+    {
```
I've refactored things here into a sensible state.

peppy committed:
Fix tolerance not being transferred to blueprint in all cases

peppy committed:
Simplify `AddOnce` call to avoid `self` argument

peppy committed:
Fix using the incorrect position for the first point

peppy committed:
Refactor b-spline path conversion code to better handle linear segments

peppy committed:
Update test coverage (and add test coverage of curve drawing)

peppy committed:
Apply NRT to `SliderPlacementBlueprint` tests

peppy committed:
Add failing test coverage of BSpline encoding parse failure

peppy committed:
Fix incorrect parsing of BSpline curve types

peppy committed:
Update framework

peppy committed:
Simplify toolbox initialisation

default0 commented on osu.Game.Rulesets.Osu/Edit/Blueprints/Sliders/SliderPlacementBlueprint.cs:
```diff
@@ -180,7 +180,7 @@ protected override bool OnDragStart(DragStartEvent e)
 
                             bSplineBuilder.Clear();
-                            bSplineBuilder.AddLinearPoint(ToLocalSpace(e.ScreenSpaceMousePosition) - HitObject.Position);
+                            bSplineBuilder.AddLinearPoint(ToLocalSpace(e.ScreenSpaceMouseDownPosition) - HitObject.Position);
```
Unsure if this fix is accurate / sensible. Shouldn't we be reading out the position of the already-placed first control point of the slider instead of relying on a reading of the mouse position at time of drag start? Mouse position of any kind should probably just be a fallback in case there is no starting control point yet.

bdach replied:
I currently can't see a circumstance wherein the two positions in questions would ever be different, but maybe you've seen something that contradicts this...?

default0 replied:
They should not be, but them not being very much relies on OnMouseDown and OnDragStart always providing the same mouse down positions. When working on this originally, there have been somewhat significant deviations on when exactly OnDragStart fired (that's why there's all this checking logic above in the first place), so it does not feel so great to me to rely on details of input event data or timing here - especially since the intention is "start the bspline builder from where the slider started" anyway. Better to express that intention directly than in this roundabout way imo.

peppy replied:
I don't quite understand this train of thought.

The previous logic was using a location that is quite far away from where the user actually mouse-downs. That means that the input points to the b-spline approximator would start from a point ~10 pixels away from the actual start of the slider, potentially leading to a less-than-optimal curve (as the approximator is missing that initial data).

The only argument I could think of is that the approximator is now receiving a "jump" in data between the first point and when the drag events start to arrive and this could throw it off? But I would see that as an issue with the approximator, not the calling code.

default0 replied:
If OnDragStarts ScreenSpaceMouseDownPosition differs from the position that OnMouseDown received to place the first control point of the slider, the starting position of the slider will be shifted to the OnDragStart position instead of remaining where the user originally placed the slider.
After messing around with it, this even causes a very noticeable bug: Place a slider (mousedown), move your cursor, start the drag at a different position.


https://github.com/ppy/osu/assets/4215321/1ee561c4-7252-4c6c-a483-ad12dd56aabe

peppy replied:
That's a regression for sure, but not related to the mouse down position. I'll fix that by bringing back the "cursor point" check I was trying to get rid of.

default0 replied:
After investigating this behaviour a bit further, it turns out there is another issue related to directly reading out the mouse position when beginning free-hand drawing. If distance snap (or, presumably, grid snap), causes the hitobjects position to deviate from the cursor position the moment OnDragStart triggers, weird artifacting (similar to the video above) starts to happen again.

I have tried to address this in #25539

peppy committed:
Remove redundant `Clear()` call

peppy committed:
Simplify drag handling code in `SliderPlacementBlueprint`

peppy committed:
Reorder methods

peppy reviewed (dismissed):
I think this is in a state I'm quite happy with now, pending resolution of open discussions.

peppy committed:
Fix dragging after one point already placed incorrectly entering drawing mode

default0 commented on osu.Game.Rulesets.Osu/Edit/Blueprints/Sliders/SliderPlacementBlueprint.cs:
```diff
@@ -171,7 +171,7 @@ protected override bool OnDragStart(DragStartEvent e)
 
             // Only enter drawing mode if no additional control points have been placed.
-            if (HitObject.Path.ControlPoints.Count > 2)
+            if (HitObject.Path.ControlPoints.Count > 2 || HitObject.Path.ControlPoints.Last() != cursor)
```
After testing with this change, I cannot start free-hand drawing at all locally anymore - presumably because the cursor cp gets placed before OnDragStart gets hit, causing this check to always be true on my end.

peppy replied:
Yeah hold your horses. I'm adding tests and fixing things.

Maybe come back in an hour or so.

peppy replied:
Should be in a good state now, at least for that particular issue. There's one more minor discrepancy I've found with the curve type choice that I'll investigate further.  

default0 commented on osu.Game.Rulesets.Osu.Tests/Editor/TestSceneSliderPlacementBlueprint.cs:
```diff
@@ -428,16 +426,16 @@ private void addClickStep(MouseButton button)
+            AddAssert($"control point {index} at {position}", () => Precision.AlmostEquals(position, getSlider()!.Path.ControlPoints[index].Position, 1));
 
-        private Slider getSlider() => HitObjectContainer.Count > 0 ? ((DrawableSlider)HitObjectContainer[0]).HitObject : null;
+        private Slider? getSlider() => HitObjectContainer.Count > 0 ? ((DrawableSlider)HitObjectContainer[0]).HitObject : null;
```
I'm really not a fan of using !. everywhere - might be better to have getSlider() be optimistic (no checks, no null return - and thus typed not nullable) and introduce a separate sliderExists() or similar for use in assertPlaced, then get rid of all the !. everywhere else?

peppy replied:
Dunno, it's used in a way where it may be null, so this is best-effort right now.

bdach replied:
It's test code, we don't really care that much. If it dies it dies.

If it was actual application code, sure.

peppy committed:
Add failing test coverage of drag after point placement

peppy committed:
Add `ToString` on `PathType` for better test output
